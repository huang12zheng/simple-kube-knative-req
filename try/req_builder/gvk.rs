use kube::api::GroupVersionKind;

pub struct GKVSpec {
    pub group: String,
    pub version: String,
    pub kind: String,
}

impl GKVSpec {
    #[allow(dead_code)]
    pub fn from_env() -> Self {
        let group = std::env::var("GROUP").unwrap_or_else(|_| "serving.knative.dev".into());
        let version = std::env::var("VERSION").unwrap_or_else(|_| "v1".into());
        let kind = std::env::var("KIND").unwrap_or_else(|_| "Service".into());
        Self {
            group,
            version,
            kind,
        }
    }
}

impl Into<GroupVersionKind> for GKVSpec {
    fn into(self) -> GroupVersionKind {
        let gkv = GroupVersionKind::gvk(&self.group, &self.version, &self.kind);
        gkv
    }
}
