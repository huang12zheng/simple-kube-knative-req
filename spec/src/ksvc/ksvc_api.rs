//! It want to get the [Api<DynamicObject>]
// ! when developing, what might come to mind:
// !
// ! # Step 1: async fn
// ! ```
// ! #[tokio::test]
// ! async fn test() {
// !     let client = CLIENT.get().await.clone();
// !
// !     let gvk: GroupVersionKind = GVKSpec::from_env().into();
// !
// !     let api: Api<DynamicObject> = ReqBuilder::default()
// !         .client(client)
// !         .gvk(gvk)
// !         .set_ar()
// !         .await
// !         .build()
// !         .into();
// !
// !     println!("{:?}", api);
// !     let str = format!("{:?}", api);
// !     let expect_str = r#"Api { request: Request { url_path: "/apis/serving.knative.dev/v1/namespaces/default/services" }, client: "...", namespace: Some("default") }"#;
// !     assert_eq!(str, expect_str);
// !
// !     let ksvc = KnativeSpec::default().ksvc();
// ! }
// ! ```
// ! # Steps 2: split async fn into multi async fn
// ! # Steps 3: make multi async fn in a trait ~~[KnativeSpecIntoReq]~~ [KnativeSpecIntoApi::get_api] with [async_trait], newtype [KnativeSpecWrapper]
// !
// ! fn in trait [KnativeSpecIntoApi] have:
// ! - get_client
// ! - get_gvk
// ! - get_ar
// ! - get_api
// !
// ! # Last Step:
// ! `api.create(&PostParams::default(), &ksvc).await?;`
// !
// ! # Other Info:
// ! - no need `get_client` in KnativeSpecIntoApi
// !
// ! make `let client =<Self as KnativeSpecIntoApi>::get_client().await;`  into
// ! `let client = CLIENT.get().await.clone();`

use async_trait::async_trait;

// #[allow(unused_imports)]
// use client::CLIENT;
// #[allow(unused_imports)]
// use gvk::*;
use kube::{
    api::{Api, ApiResource},
    core::{DynamicObject, GroupVersionKind},
    Client,
};

use crate::gvk::IntoGVKSpec;

use super::{client::CLIENT, ksvc::KnativeSpec};

// !due to KnativeSpecWrapper, there is no need Req
// !```
// !struct Req {
// !    pub client: Client,
// !    pub ar: ApiResource,
// !}
// !impl Into<Api<DynamicObject>> for Req {
// !    fn into(self) -> Api<DynamicObject> {
// !        let api = Api::<DynamicObject>::default_namespaced_with(self.client.to_owned(), &self.ar);
// !        api
// !    }
// !}
// !```
/// newtype wrapper for [KnativeSpec]
#[derive(Debug)]
pub struct KnativeSpecWrapper(pub KnativeSpec);
#[async_trait]
/// trait for [KnativeSpecWrapper] to get [Api]
pub trait KnativeSpecIntoApi {
    /// merge [Client] and [ApiResource] into [Api]
    async fn get_api(&self) -> Api<DynamicObject>;
    /// Get [ApiResource] from [GroupVersionKind]
    fn get_ar(&self) -> ApiResource;

    /// Get [Client] from [CLIENT]
    async fn get_client(&self) -> &'async_trait Client {
        return CLIENT.get().await;
    }
    /// Get gvk from KnativeSpec with trait [IntoGVKSpec]
    fn get_gvk(&self) -> GroupVersionKind;
}

#[async_trait]
impl KnativeSpecIntoApi for KnativeSpecWrapper {
    async fn get_api(&self) -> Api<DynamicObject> {
        let client = self.get_client().await.clone();
        let ar = self.get_ar();
        let api = Api::<DynamicObject>::default_namespaced_with(client, &ar);
        api
    }

    // !20220713 hz
    // !instead of use async fn, ApiResource::from_gvk(&gvk) do better
    // !async fn get_ar(&self) -> ApiResource {
    // !    let client = CLIENT.get().await;
    // !    let gvk = self.get_gvk();
    // !    let ar = discovery::pinned_kind(client, &gvk).await;
    // !    // insta::assert_debug_snapshot!(ar);
    // !    let (ar, _) = discovery::pinned_kind(client, &gvk).await.unwrap();
    // !    ar
    // !}
    fn get_ar(&self) -> ApiResource {
        return ApiResource::from_gvk(&self.get_gvk());
    }

    fn get_gvk(&self) -> GroupVersionKind {
        let gvk = <KnativeSpec as IntoGVKSpec>::into_gvk(&self.0);
        return gvk.into();
    }
}

#[cfg(test)]
#[tokio::test]
async fn show_ar() {
    use envconfig::Envconfig;
    let gvk: GroupVersionKind = crate::gvk::GVKSpec::init_from_env().unwrap().into();

    let ar = ApiResource::from_gvk(&gvk);
    insta::assert_debug_snapshot!(ar);
}
#[cfg(test)]
#[tokio::test]
async fn show_api() {
    use envconfig::Envconfig;
    let client: Client = CLIENT.get().await.clone();
    let gvk: GroupVersionKind = crate::gvk::GVKSpec::init_from_env().unwrap().into();
    let ar = ApiResource::from_gvk(&gvk);
    let api = Api::<DynamicObject>::default_namespaced_with(client, &ar);
    insta::assert_debug_snapshot!(api);
}
