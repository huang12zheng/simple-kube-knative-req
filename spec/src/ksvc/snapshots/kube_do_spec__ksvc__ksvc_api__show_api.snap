---
source: spec/src/ksvc/ksvc_api.rs
expression: api
---
Api {
    request: Request {
        url_path: "/apis/serving.knative.dev/v1/namespaces/default/services",
    },
    client: "...",
    namespace: Some(
        "default",
    ),
}
